File: data-processor/telemetry/otel.py
import os
import logging
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

logger = logging.getLogger(__name__)


def init_tracer():
    resource = Resource.create({"service.name": "data-processor"})
    provider = TracerProvider(resource=resource)

    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317")

    try:
        # Export to remote OTLP collector endpoint (Grafana/Jaeger)
        otlp_exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
        processor = BatchSpanProcessor(otlp_exporter)
        provider.add_span_processor(processor)
    except Exception as e:
        logger.warning(
            f"Failed to initialize OTLP exporter: {e}. Falling back to console."
        )
        console_processor = BatchSpanProcessor(ConsoleSpanExporter())
        provider.add_span_processor(console_processor)

    trace.set_tracer_provider(provider)
    return trace.get_tracer(__name__)

File: ingestion-api/telemetry/otel.go
package telemetry

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
)

// InitTracer initializes an OTLP exporter, and configures the corresponding trace and
// metric providers. If the OTEL Collector is unreachable, it returns a no-op provider
// so the API can still start without tracing.
func InitTracer() (*sdktrace.TracerProvider, error) {
	// Use a short timeout so the API doesn't hang if the collector is down.
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		endpoint = "localhost:4317" // Default Jaeger/Collector
	}

	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithInsecure(),
		otlptracegrpc.WithEndpoint(endpoint),
	)
	if err != nil {
		log.Printf("[telemetry] WARN: could not create OTLP exporter (%v) — tracing disabled", err)
		return newNoopProvider(), nil
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName("ingestion-api"),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(tp)
	log.Printf("[telemetry] OpenTelemetry tracer initialized → exporting to %s", endpoint)
	return tp, nil
}

// newNoopProvider returns a TracerProvider that records spans but drops them.
// This lets the middleware run without errors even when the collector is offline.
func newNoopProvider() *sdktrace.TracerProvider {
	tp := sdktrace.NewTracerProvider() // no exporter = spans are dropped
	otel.SetTracerProvider(tp)
	return tp
}
